### PR TITLE
[Build]: Support to use symbol links for lazy installation targets to reduce the image size

### PR DESCRIPTION
**Pull request by xumia**
#### Why I did it
Support to use symbol links in platform folder to reduce the image size.
The current solution is to copy each lazy installation targets (xxx.deb files) to each of the folders in the platform folder. The size will keep growing when more and more packages added in the platform folder. For cisco-8000 as an example, the size will be up to 2G, while most of them are duplicate packages in the platform folder.
 
#### How I did it
Create a new folder in platform/common, all the deb packages are copied to the folder, any other folders where use the packages are the symbol links to the common folder.

Why platform.tar?
We have implemented a patch for it, see https://github.com/Azure/sonic-buildimage/pull/10775, but the problem is the the onie use really old unzip version, cannot support the symbol links.
The current solution is similar to the PR 10775, but make the platform folder into a tar package, which can be supported by onie. During the installation, the package.tar will be extracted to the original folder and removed.

#### How to verify it
1. Verified Arista installation
Installed by sonic-installer.
```
+ unzip -oq /tmp/tmp.0QZcH4zcM2/sonic_image.swi -x boot0 dockerfs.tar.gz platform.tar.gz -d /host/image-master-10923.107767-68d3228d0
+ info Extracting platform.tar.gz
+ cut -f1 -d  /proc/uptime
+ printf %04.2f: Extracting platform.tar.gz\n 28172.20
+ mkdir -p /host/image-master-10923.107767-68d3228d0/platform
+ tar xzf - -C /host/image-master-10923.107767-68d3228d0/platform
+ unzip -oqp /tmp/tmp.0QZcH4zcM2/sonic_image.swi platform.tar.gz
+ grep  /host  /proc/mounts

root@str-a7060cx-acs-1:/host/image-master-10923.107767-68d3228d0# ls
allowlist_paths.conf  boot  docker  fs.squashfs  kernel-cmdline  platform  sonic-config  swi-signature  version
root@str-a7060cx-acs-1:/host/image-master-10923.107767-68d3228d0# find platform/
platform/
platform/x86_64-grub
platform/x86_64-grub/grub-pc-bin_2.04-20_amd64.deb
platform/firsttime
```
Boot from Aboot.
```
Aboot# boot flash:sonic.swi
202.64: Cleaning flash content /mnt/flash
211.48: Generating boot-config, machine.conf and cmdline
211.59: Installing image under /mnt/flash/image-master-10923.107767-68d3228d0
211.60: Moving swi to a tmpfs
213.50: Extracting swi content
215.34: Extracting platform.tar.gz
215.38: Extracting dockerfs.tar.gz from swi

root@sonic:/host# ls
acl.json     image-master-10923.107767-68d3228d0  machine.conf  warmboot
boot-config  kernel-params-base                   reboot-cause
root@sonic:/host# ls image-master-10923.107767-68d3228d0/
allowlist_paths.conf  docker       kernel-cmdline  rw             version
boot                  fs.squashfs  platform        swi-signature  work
root@sonic:/host# ls image-master-10923.107767-68d3228d0/platform/
x86_64-grub
root@sonic:/host# 

```

2. Verified Mellanox, installed by onie
```
Installing SONiC to /tmp/tmp.dm5r6G/image-master-10923.107767-68d3228d0
Archive:  fs.zip
   creating: boot/
  inflating: boot/config-5.10.0-12-2-amd64
  inflating: boot/initrd.img-5.10.0-12-2-amd64
  inflating: boot/vmlinuz-5.10.0-12-2-amd64
  inflating: boot/System.map-5.10.0-12-2-amd64
  inflating: fs.squashfs
Success: Support tarball created: /tmp/onie-support-mlnx_msn2700.tar.bz2
Installing for i386-pc platform.
Installation finished. No error reported.
Switch CPU vendor is: GenuineIntel
Switch CPU cstates are: disabled
EXTRA_CMDLINE_LINUX=sonic_fips=1
Installed SONiC base image SONiC-OS successfully


root@sonic:/host/image-master-10923.107767-68d3228d0# ls platform/
x86_64-grub

```

3. Verified on cisco-8000, installed by onie, it worked fine.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

**Comment by lguohan** [comment]
please add description, why you are doing this?

**Comment by xumia** [comment]
> please add description, why you are doing this?

@lguohan , update the description.

File: files/build_templates/sonic_debian_extension.j2 [hunk 1]
```diff
@@ -610,7 +610,9 @@ sudo LANG=C chroot $FILESYSTEM_ROOT depmod -a {{kversion}}
 sudo dpkg --root=$FILESYSTEM_ROOT -i {{deb}} || sudo LANG=C DEBIAN_FRONTEND=noninteractive chroot $FILESYSTEM_ROOT apt-get -y install -f --download-only
 
 sudo mkdir -p $FILESYSTEM_ROOT/$PLATFORM_DIR/{{dev}}
-sudo cp {{ deb }} $FILESYSTEM_ROOT/$PLATFORM_DIR/{{dev}}/
+sudo mkdir -p $FILESYSTEM_ROOT/$PLATFORM_DIR/common
+sudo cp {{ deb }} $FILESYSTEM_ROOT/$PLATFORM_DIR/common/
+sudo ln -sf "../common/{{ debfilename }}" "$FILESYSTEM_ROOT/$PLATFORM_DIR/{{dev}}/{{ debfilename }}"
```
Comment (lguohan) [inline files/build_templates/sonic_debian_extension.j2:615]: why don't we do the same thing (symbol link approach for deb in /var/cache/apt/archives?

**Comment by lguohan** [comment]
i am wondering what is the saving for sonic-broadcom.bin?

**Comment by xumia** [comment]
> 



> i am wondering what is the saving for sonic-broadcom.bin?

There is a zip file "fs.zip" in installer folder in the .bin file. In the fs.zip, it contains boot/, platform/, docker.tar.gz, and fs.squashfs.
There are some of the .deb files in platform folders, we change to use symbol links, because of multiple duplicate files, to reduce the image size.
We change the platform/ to platform.tar.gz, since the unzip program in ONIE does not support to extract the symbol links in .zip file (fs.zip). During installing, the platform.tar.gz is changed back to platform/.

File: files/build_templates/sonic_debian_extension.j2 (hunk 1 above)
Comment (xumia) [reply files/build_templates/sonic_debian_extension.j2:615]: It is to copy the packages to installer folder. The packages will be installed by the script (install.sh) by each of the boot loader. In the platform folder, it contains some sub folders, such as x86_64-8101_32fh_o-r0, x86_64-grub, the symbol links are created in the sub folders.
@lguohan , do you mean to copy the packages into /var/cache/apt/archives? The platform folder is created before the filesystem ready, maybe the folder can not be used.

**Comment by xumia** [comment]
@lguohan , could you please help review it?

**Comment by xumia** [comment]
@Staphylo, could you please help review the PR? Thanks.

**Comment by abdosi** [comment]
cc @anamehra

File: files/Aboot/boot0.j2 [hunk 2]
```diff
@@ -350,6 +350,12 @@ extract_image() {
     ## Unzip the image except boot0 and dockerfs archive
     unzip -oq "$swipath" -x boot0 "$dockerfs" -d "$image_path"
 
+    ## Extrat the platform.tar.gz
```
[the remark this answers is not in the document]
Comment (xumia) [reply files/Aboot/boot0.j2]: Fixed, thanks.

File: files/Aboot/boot0.j2 [hunk 3]
```diff
@@ -350,6 +350,12 @@ extract_image() {
...
+    info "Extracting platform.tar.gz"
+    mkdir -p $image_path/platform
```
[the remark this answers is not in the document]
Comment (xumia) [reply files/Aboot/boot0.j2]: Fixed.

File: files/Aboot/boot0.j2 [hunk 4]
```diff
@@ -350,6 +350,12 @@ extract_image() {
...
+    tar xzf $image_path/platform.tar.gz -C $image_path/platform
```
[the remark this answers is not in the document]
Comment (xumia) [reply files/Aboot/boot0.j2]: Thanks for the comment, changed to unzip to pipe, and extract in memory.

[the remark this answers is not in the document]
Comment (Staphylo) [reply files/Aboot/boot0.j2]: LGTM, thanks for addressing this

**Comment by jon-nokia** [comment]
202012 branch is waiting for either this PR to be merged or we need the Revert "Reduce image size for lazy installation packages (#10775)" #10916 Please advise.

**Comment by xumia** [comment]
@lguohan , @Staphylo , do you have more comments? If not, could you please help approve it? Thanks.

**Review by Staphylo — APPROVED** [review]
LGTM on the platform independent code and Aboot boot0

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**